src/data-model-providers/codedriven/CodeDrivenDataModelProvider.cpp

```diff
@@ -0,0 +1,337 @@
+#include <protocols/interaction_model/StatusCode.h>
+
+using chip::Protocols::InteractionModel::Status;
+namespace chip {
+namespace app {
+
+CHIP_ERROR CodeDrivenDataModelProvider::Startup(DataModel::InteractionModelContext context)
+{
+    ReturnErrorOnFailure(DataModel::Provider::Startup(context));
+
+    mServerClusterContext.reset();
+    mServerClusterContext.emplace(ServerClusterContext({
+        .provider           = this,
+        .storage            = mPersistentStorageDelegate,
+        .interactionContext = &mContext,
+    }));
+
+    // Startup all server clusters across all endpoints
+    bool had_failure = false;
+    for (auto * endpointProvider : mEndpointProviderRegistry)
+    {
+        ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+        ReturnErrorOnFailure(endpointProvider->ServerClusterInterfaces(serverClusterBuilder));
+        auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+        for (auto * serverCluster : serverClusters)
+
+        {
+            CHIP_ERROR err = serverCluster->Startup(*mServerClusterContext);
+            if (err != CHIP_NO_ERROR)
+            {
+                had_failure = true;
+                VerifyOrReturnError(!serverCluster->GetPaths().empty(), CHIP_ERROR_INCORRECT_STATE);
+                // Log startup failure for the specific cluster
+                const ConcreteClusterPath path = serverCluster->GetPaths().front();
+                ChipLogError(DataManagement, "Cluster " ChipLogFormatMEI " on Endpoint %u startup failed: %" CHIP_ERROR_FORMAT,
+                             ChipLogValueMEI(path.mClusterId), path.mEndpointId, err.Format());
+            }
+        }
+    }
+    if (had_failure)
+    {
+        return CHIP_ERROR_HAD_FAILURES;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CodeDrivenDataModelProvider::Shutdown()
+{
+    // Shutdown all server clusters across all endpoints
+    for (auto * endpointProvider : mEndpointProviderRegistry)
+    {
+        ReadOnlyBufferBuilder<ServerClusterInterface *> serverClusterBuilder;
+        ReturnErrorOnFailure(endpointProvider->ServerClusterInterfaces(serverClusterBuilder));
+        auto serverClusters = serverClusterBuilder.TakeBuffer();
+
+        for (auto * serverCluster : serverClusters)
+        {
+            // TODO: The API for serverCluster->Shutdown is swallowing errors here. We should update it to return CHIP_ERROR.
+            serverCluster->Shutdown();
+        }
+    }
+    mServerClusterContext.reset();
+    return CHIP_NO_ERROR;
+}
+
+DataModel::ActionReturnStatus CodeDrivenDataModelProvider::ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                                         AttributeValueEncoder & encoder)
+{
+    EndpointProviderInterface * epProvider = GetEndpointProvider(request.path.mEndpointId);
+    if (epProvider == nullptr)
+    {
+        return Status::UnsupportedEndpoint;
+    }
```

Comment:
```suggestion
    VerifyOrReturnError(epProvider != nullptr, Status::UnsupportedEndpoint);
```

nitpicky throughout: would the verify macros be more common across matter SDK code? The if is just as readable thogh so just offering this as an option. Feel free to discard based on your style choice.